forc-pkg/src/pkg.rs

```diff
@@ -2044,6 +2058,14 @@ fn validate_contract_deps(graph: &Graph) -> Result<()> {
     Ok(())
 }
 
+/// Represents the result of building a package.
+///
+/// For non-contract packages second element of the tuple will always be `None`.
+///
+/// For contract packages second element of the tuple will hold the contract re-compiled without
+/// thte tests included.
+type BuildResult = (BuiltPackage, Option<BuiltPackage>);
```

Comment:
Hmmm I'm not sure `BuildResult` is a good name for this. `Result` is normally used for result types, i.e. types that might be an error.

Perhaps we're better off splitting up `BuiltPackage` into something like the following?

```rust
/// The result of successfully compiling a package.
#[derive(Debug, Clone)]
pub struct BuiltPackage {
    pub name: String,
    pub descriptor: BuiltPackageDescriptor,
    pub tree_type: TreeType,
    pub build_target: BuildTarget,
    pub json_abi_program: ProgramABI,
    pub storage_slots: Vec<StorageSlot>,
    source_map: SourceMap,
    pub bytecode: BuiltPackageBytecode,
    /// `Some` for contract member builds where tests were included. This is
    /// required so that we can deploy once instance of the contract (without
    /// tests) with a valid contract ID before executing the tests as scripts.
    ///
    /// For non-contract members, this is always `None`.
    pub bytecode_without_tests: Option<BuiltPackageBytecode>,
}

/// The bytecode associated with a built package along with its entry points.
struct BuiltPackageBytecode {
    pub data: Vec<u8>,
    pub entries: Vec<PkgEntry>,
}
```

This also helps to avoid duplicating stuff that stays the same (name, descriptor, etc).